File: gallery/process_counter_android/app/ui/CounterBoardUiState.kt
/** Snapshot for Compose / RecyclerView binding (imaginary UI). */
data class RowUiState(
    val processId: Int,
    val label: String,
    val reps: Int,
    val ticks: Int,
    val running: Boolean,
    val selected: Boolean,
)

data class CounterBoardUiState(
    val title: String,
    val processPath: String,
    val processId: Int,
    val stateGeneration: Int,
    val rows: List<RowUiState>,
    val sumValues: Int,
)

fun CounterBoardPage.toUiState(): CounterBoardUiState {
    val n = rowCount()
    val sel = selectedIndex
    val rowStates = mutableListOf<RowUiState>()
    for (i in 0 until n) {
        val r = rows[i]
        rowStates.add(
            RowUiState(
                processId = r.__rangerId,
                label = r.label,
                reps = r.value,
                ticks = r.tickCount,
                running = r.running,
                selected = i == sel,
            )
        )
    }
    return CounterBoardUiState(
        title = title,
        processPath = __rangerPath,
        processId = __rangerId,
        stateGeneration = __rangerStateGeneration,
        rows = rowStates,
        sumValues = sumValues(),
    )
}

File: gallery/process_counter_ios/ios/UI/CounterBoardUiState.swift
struct RowUiState: Identifiable {
    var id: Int { processId }
    let processId: Int
    let label: String
    let reps: Int
    let ticks: Int
    let running: Bool
    let selected: Bool
}

struct CounterBoardUiState {
    let title: String
    let processPath: String
    let processId: Int
    let stateGeneration: Int
    let rows: [RowUiState]
    let sumValues: Int
}

extension CounterBoardPage {
    func toUiState() -> CounterBoardUiState {
        let n = rowCount()
        var rowStates: [RowUiState] = []
        for i in 0..<n {
            let r = rows[i]
            rowStates.append(
                RowUiState(
                    processId: r.__rangerId,
                    label: r.label,
                    reps: r.value,
                    ticks: r.tickCount,
                    running: r.running,
                    selected: i == selectedIndex
                )
            )
        }
        return CounterBoardUiState(
            title: title,
            processPath: __rangerPath,
            processId: __rangerId,
            stateGeneration: __rangerStateGeneration,
            rows: rowStates,
            sumValues: sumValues()
        )
    }
}
